src/protocol/mod.rs

```diff
@@ -7,27 +7,82 @@ pub mod sender;
 /// Timestamp in us
 pub type TimeStamp = i32;
 
+/// Duration in us, e.g. RTT
+pub type TimeSpan = i32;
+
 #[derive(Copy, Clone, Debug)]
 pub struct TimeBase(Instant);
 impl TimeBase {
-    pub fn new() -> Self {
-        Self(Instant::now())
-    }
-
-    pub fn from_raw(start_time: Instant) -> Self {
-        let _ = Self::new();
+    pub fn new(start_time: Instant) -> Self {
         Self(start_time)
     }
 
     #[allow(clippy::trivially_copy_pass_by_ref)]
     pub fn timestamp_from(&self, instant: Instant) -> TimeStamp {
-        let elapsed = instant - self.0;
-        elapsed.as_micros() as i32 // TODO: handle overflow here
+        if self.0 > instant {
+            0 - (self.0 - instant).as_micros() as i32
+        } else {
+            (instant - self.0).as_micros() as i32
+        }
     }
 
     #[allow(clippy::trivially_copy_pass_by_ref)]
     pub fn instant_from(&self, timestamp: TimeStamp) -> Instant {
-        self.0 + Duration::from_micros(timestamp as u64)
+        if timestamp < 0 {
+            self.0 - Duration::from_micros(timestamp.abs() as u64)
+        } else {
+            self.0 + Duration::from_micros(timestamp as u64)
+        }
+    }
+
+    pub fn adjust(&mut self, delta: TimeSpan) {
+        match delta {
+            delta if delta > 0 => {
+                self.0 += Duration::from_micros(delta as u64);
+            }
+            delta if delta < 0 => {
+                self.0 -= Duration::from_micros(delta.abs() as u64);
+            }
+            _ => {}
+        }
+    }
+}
+
+#[cfg(test)]
+mod timebase {
```

Comment:
I've been calling the test module `test`

Reply:
Will this work if I locate unit test modules near their corresponding "units" (e.g. structs, enums, functions, etc.). Rust modules are a bit heavy weight, and I would prefer to have larger module files that contain many "units", and having a unit test module per "unit" like this allowed me to easily isolate and run all the tests for a given "unit" without having to create a separate module per "unit". The naming of these modules is a secondary concern to me.

Reply:
Yeah, if there's a good reason then I'm fine with it.

I'm willing to discuss, but I generally prefer smaller files. I also so far prefer just having the unit tests at the bottom so it doesn't clutter the code, but I'm pretty new to rust and my preferences are still evolving :).